src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java

```diff
@@ -0,0 +1,708 @@
+            this.studyByNameNoScope.put(importRow.getEnv(), studyPIO);
+
+            PendingImportObject<BrAPIObservationUnit> obsUnitPIO = createObsUnitPIO(program, commit, obsUnitNextVal, importRow);
+            String key = createObservationUnitKey(importRow);
+            this.observationUnitByNameNoScope.put(key, obsUnitPIO);
+        }
+    }
+
+    private String createObservationUnitKey(ExperimentObservation importRow) {
+        String key = importRow.getEnv() + importRow.getExpUnitId();
+        return key;
+    }
+
+    private ValidationErrors validateFields(List<BrAPIImport> importRows, ValidationErrors validationErrors) {
+        HashSet<String> uniqueStudyAndObsUnit = new HashSet<>();
+        for (int i = 0; i < importRows.size(); i++) {
+            ExperimentObservation importRow = (ExperimentObservation) importRows.get(i);
+            validateConditionallyRequired(validationErrors, i, importRow);
+            validateUniqueObsUnits(validationErrors, uniqueStudyAndObsUnit, i, importRow);
+        }
+        return validationErrors;
+    }
+
+    /**
+     * Validate that the the observation unit is unique within a study.
+     * <br>
+     * SIDE EFFECTS:  validationErrors and uniqueStudyAndObsUnit can be modified.
+     * @param validationErrors can be modified as a side effect.
+     * @param uniqueStudyAndObsUnit can be modified as a side effect.
+     * @param i counter that is always two less the file row being validated
+     * @param importRow the data row being validated
+     */
+    private void validateUniqueObsUnits(
+            ValidationErrors validationErrors,
+            HashSet<String> uniqueStudyAndObsUnit,
+            int i,
+            ExperimentObservation importRow) {
+        String envIdPlusStudyId = createObservationUnitKey( importRow );
+        if( uniqueStudyAndObsUnit.contains( envIdPlusStudyId )){
+            String errorMessage = String.format("The ID (%s) is not unique within the environment(%s)", importRow.getExpUnitId(), importRow.getEnv());
+            this.addRowError("Exp Unit ID", errorMessage, validationErrors, i);
+        }
+        else{
+            uniqueStudyAndObsUnit.add( envIdPlusStudyId );
+        }
+    }
+
+    private void validateConditionallyRequired(ValidationErrors validationErrors, int i, ExperimentObservation importRow) {
+        String experimentTitle = importRow.getExpTitle();
+        String obsUnitID = importRow.getObsUnitID();
+        if( StringUtils.isBlank( obsUnitID )){
+            validateRequiredCell(
+                    experimentTitle,
+                    "Exp Title",
+                    "Field is blank", validationErrors, i
+                    );
+        }
+
+        ImportObjectState expState = this.trialByNameNoScope.get(experimentTitle).getState();
+        boolean isExperimentNew = (expState == ImportObjectState.NEW);
+
+        if (isExperimentNew) {
```

Comment:
What about the case that the trial exists and the study exists and you are just trying to add obs units to a study. Seems like you would need two checks.

If trial name is new:
- Trial data is required
- Study data is required
- Obs Unit data is required

If Env is new:
- Env data is required
- ObsUnitData is required

Reply:
@shawnyarnes please correct what I may get wrong.

The experiment upload functionality should not allow you to append observation units to an existing environment (aka study).  Observation Units can only be added to an environment when the environment is first created.